symbols: support group references in replacements

sthibaul opened the pull request:
### Link to issue number:

fixes #11107

### Summary of the issue:

Rules such as `dates .` are quite complex to write: `((?<=\b\d\d)\.(?=\d\d.(\d{2}|\d{4})\b))|((?<=\b\d\d.\d\d)\.(?=(\d{2}|\d{4})\b))` because one wants to change two pieces of text here, and so we have to match them twice. Other rules are even more complex to write (e.g. `1,20€` -> `1 euro 20` needs two complex symbols to perform the two differing `,`->` euro ` `€`->`` replacements.

### Description of how this pull request fixes the issue:

This adds support for regexp group references.

### Testing performed:

I tested the use of it in the French translation, included in the commit:

```
dates .        \b(\d\d)\.(\d\d)\.(\d{2}|\d{4})\b
dates /        \b(\d\d)/(\d\d)/(\d{2}|\d{4})\b
```
```
dates .        \1 point \2 point \3        all     always  # point de date
dates /        \1 barre oblique \2 barre oblique \3        all     always  # barre oblique de date
```

The change uses Python's standard support for group, already used for the complex group parsing.

### Known issues with pull request:

It requires checking that no symbols file includes a backslash in the replacement string. I did this, and found two (commented) occurrences in the italian symbols file, fixed by the commit. (and incidently found a typo in the hebrew file)

### Change log entry:

*New features*
`Symbol pronunciation: Support grouping in complex symbol definition and group references in replacement rule, making them more powerful and simpler. (#11107)`

sthibaul committed:
symbols: support group references in replacements

Fixes #11107

josephsl commented:
Hi, this is more towards changes for developers, although it would be beneficial to talk about it briefly as part of the user guide if warranted. Thanks.

 

From: Samuel Thibault <notifications@github.com> 
Sent: Tuesday, May 5, 2020 3:02 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [nvaccess/nvda] symbols: support group references in replacements (#11116)

 


Link to issue number:


fixes #11107 <https://github.com/nvaccess/nvda/issues/11107> 


Summary of the issue:


Rules such as dates . are quite complex to write: ((?<=\b\d\d)\.(?=\d\d.(\d{2}|\d{4})\b))|((?<=\b\d\d.\d\d)\.(?=(\d{2}|\d{4})\b)) because one wants to change two pieces of text here, and so we have to match them twice. Other rules are even more complex to write (e.g. 1,20€ -> 1 euro 20 needs two complex symbols to perform the two differing ,->€ €`->`` replacements)


Description of how this pull request fixes the issue:


This adds support for regexp group references.


Testing performed:


I tested the use of it in the French translation, included in the commit:

dates .        \b(\d\d)\.(\d\d)\.(\d{2}|\d{4})\b
 
dates /        \b(\d\d)/(\d\d)/(\d{2}|\d{4})\b
 
dates .        \1 \2 \3        all     always  # point de date
 
dates /        \1 \2 \3        all     always  # barre oblique de date
 

The change uses Python's standard support for group, already used for the complex group parsing.


Known issues with pull request:


It requires checking that no symbols file includes a backslash in the replacement string. I did this, and found two (commented) occurrences in the italian symbols file (and incidently found a typo in the hebrew file)


Change log entry:


New features
symbols: Support group references in replacements. (#11107)

  _____  


You can view, comment on, or merge this pull request online at:


  https://github.com/nvaccess/nvda/pull/11116


Commit Summary


*	symbols: support group references in replacements


File Changes


*	M source/characterProcessing.py <https://github.com/nvaccess/nvda/pull/11116/files#diff-aa66a136e3a6d4384d75b3efaec3c8f4>  (21) 
*	M source/locale/fr/symbols.dic <https://github.com/nvaccess/nvda/pull/11116/files#diff-603f2f0f62c5294f61d8307a69dc7d6c>  (8) 
*	M source/locale/he/symbols.dic <https://github.com/nvaccess/nvda/pull/11116/files#diff-bd71c6d92261e13912e9b575a5e90d40>  (2) 
*	M source/locale/it/symbols.dic <https://github.com/nvaccess/nvda/pull/11116/files#diff-7b8a59358203ed888b1cee2a857ff89e>  (4) 


Patch Links:


*	https://github.com/nvaccess/nvda/pull/11116.patch
*	https://github.com/nvaccess/nvda/pull/11116.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11116> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEC7G56MF6PSF6X263LRQCEHJANCNFSM4MZ6EN7A> .

sthibaul commented:
> Hi, this is more towards changes for developers, 

Do you mean that I shouldn't have submitted the pull request here, but rather to another repository?

josephsl commented:
Hi, no, my comment relates to changes entry. Thanks.

 

From: Samuel Thibault <notifications@github.com> 
Sent: Tuesday, May 5, 2020 3:09 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Comment <comment@noreply.github.com>
Subject: Re: [nvaccess/nvda] symbols: support group references in replacements (#11116)

 

Hi, this is more towards changes for developers,

Do you mean that I shouldn't have submitted the pull request here, but rather to another repository?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11116#issuecomment-624332638> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEDOE44G4QKJKQ7S723RQCE6FANCNFSM4MZ6EN7A> .

sthibaul commented:
I added some words in the ./devDocs/developerGuide.t2t

sthibaul commented:
> Hi, no, my comment relates to changes entry. 

Ah, ok, I reworded it

sthibaul commented:
> Hi, no, my comment relates to changes entry.

Maybe https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples#change-log-entry should say it explicitly that the text should be meant for users?

Andre9642 commented:
Hi Samuel :)
It seems to me that modifications made in 'locale/{fr,he,it}/symbols.dic' should be done via Assembla exclusively.
Would you like to join the team of translators?

sthibaul commented:
Heya André :)
I can drop these from the MR, none of them actually have any effect, it was more for documentation.
I don't really plan to join the translators team, I just meant to pass on at least the date example so people get the idea and can then take benefit from it.

feerrenrut commented:
@josephsl can you comment on whether it will cause problems for the current translation system if changes are accepted to languages other than English?

josephsl commented:
Hi, as translations come from SVN and are done by individual translators, it would be best to push to English symbols file, which will then pass onto translators as a diff later. Thanks,

 

From: Reef Turner <notifications@github.com> 
Sent: Thursday, May 7, 2020 12:32 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] symbols: support group references in replacements (#11116)

 

@josephsl <https://github.com/josephsl>  can you comment on whether it will cause problems for the current translation system if changes are accepted to languages other than English?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11116#issuecomment-625082446> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEFESPWS4Z36BCAUUL3RQJPW7ANCNFSM4MZ6EN7A> .

sthibaul commented:
Here there are no english changes, just fixes in a couple of translations, and an improvement in another one.

sthibaul commented:
To make things simpler, I have now dropped the locale/ changes, and we can commit the locale/ fixes separately

Adriani90 commented:
Will this also fix issue #2676?

sthibaul commented:
> Will this also fix issue #2676?

No, it's unrelated, the grouping mentioned here is not of the same kind as in #2676

sthibaul commented:
So, any comment on including this?

I would like to include the same support in speech-dispatcher which uses the same dict files, but I would rather avoid making the file format diverge...

sthibaul commented:
Hello, any comment on this?

feerrenrut commented on source/characterProcessing.py:
```diff
@@ -540,15 +540,34 @@ def _regexpRepl(self, m):
 				# Complex symbol.
 				index = int(group[1:])
 				symbol = self._computedComplexSymbolsList[index]
+				replacement = ''
+
+				escape = False
+				for c in symbol.replacement:
+					if not escape:
+						if c == '\\':
+							escape = True
+						else:
+							replacement += c
+					else:
+						if c == '\\':
+							replacement += '\\'
+						elif c >= '0' and c <= '9':
+							replacement += m.group(m.lastindex + ord(c) - ord('0'))
+						else:
+							log.error("Invalid reference \\%s" % c)
+						escape = False
```
Please extract this block to a named function. Did you look to see if the regex module can handle these escapes directly? I'm concerned this approach will be error prone. If this must be done manually, please add unit tests.

sthibaul replied:
Ah, let me check, it seems like m.expand could do it.

sthibaul replied:
No, it doesn't work, because the numbers of the references that are used by `m.expand` are relative to the whole regexp, and not relative to the regexp inside the named group.

sthibaul replied:
(i.e. it's the `m+lastindex + ` here which is missing in the `re`-provided implementation)

sthibaul replied:
Is the dic engine actually tested at all? I don't find anything existing in the repository.

sthibaul replied:
I have added a test, but I don't know why the python linter doesn't like the heading comments, which are just exactly like in other files.

(there is also a fix in the french locale file, I have contacted the translator to get it fixed ahead of this)

lukaszgo1 replied:
Commends in other files were mostly introduced before we introduced the linter unfortunately. You need to insert space after each hash sign to make it happy and add one more blank line before 
`class TestComplex(unittest.TestCase):`

Also were no longer placing the file name as the first line in the copyright header so please remove it.

sthibaul replied:
Ah, the linter only looks at new code, that's why. Ok, pushed a fixed version, thanks for the explanation.

feerrenrut reviewed (changes requested):
Thanks for the contribution @sthibaul and apologies for the delay to review the work.

sthibaul commented on devDocs/developerGuide.t2t:
```diff
@@ -97,7 +98,7 @@ Certain characters cannot be typed into the file, so the following special seque
  - \f: form feed
  - \#: # character (needed because # at the start of a line denotes a comment)
-- replacement: The text which should be spoken for the symbol.
+- replacement: The text which should be spoken for the symbol. If the symbol is a complex symbol, \1, \2, etc. can be used to refer to the groups matches, which will be inlined in the replacement, allowing for simpler rules. This also means that to get a \ character in the replacement, one has to type \\.
```
Right, done so.

sthibaul commented on source/characterProcessing.py:
```diff
@@ -540,15 +540,34 @@ def _regexpRepl(self, m):
+				escape = False
+				for c in symbol.replacement:
+					if not escape:
+						if c == '\\':
```
Unfortunately `\` seems to be still interpreted in raw strings:

```
>>> r'\'
  File "<stdin>", line 1
    r'\'
       ^
SyntaxError: EOL while scanning string literal
```

feerrenrut commented on tests/unit/test_characterProcessing.py:
```diff
@@ -0,0 +1,21 @@
+	"""Test the complex symbols rules.
+	"""
+
+	def test_group(self):
```
This is a good start! Thanks for doing it. Apologies if you are already planning on expanding this, but I really want to also see tests for the `_replaceGroups` method. That's the part that contains the custom code. Please try to ensure each branch is hit with the tests.

sthibaul replied:
The test_group function was already triggering a call to `_replaceGroups`. But that wasn't exercising the '\\' case in group replacement so I have added a second test.

feerrenrut commented on source/characterProcessing.py:
```diff
@@ -519,6 +519,26 @@ def __init__(self, locale):
+		escape = False
+		for c in s:
+			if not escape:
+				if c == '\\':
```
Is there a reason not to use a raw string here (to remove confusion around the number of slash characters required)? 
```suggestion
				if c == r'\':

```

sthibaul replied:
This was already raised in another comment: that doesn't work, \ is still interpreted in raw strings:

>>> r'\'
  File "<stdin>", line 1
    r'\'
       ^
SyntaxError: EOL while scanning string literal

feerrenrut replied:
Ah. Apologies.

feerrenrut commented on source/characterProcessing.py:
```diff
@@ -519,6 +519,26 @@ def __init__(self, locale):
 			raise LookupError
 
+	def _replaceGroups(self, m, s):
+		res = ''
```
Please add type hints and a doc string for this function. Generally we prefer more descriptive variable names.

sthibaul replied:
Ok, done so

feerrenrut commented on tests/unit/test_characterProcessing.py:
```diff
@@ -0,0 +1,39 @@
+
+	def _replace(self, m):
+		group = m.lastgroup
+		if group == 'date':
```
Why is this specific to date?

sthibaul replied:
It is just the name of the group below, I could have put `bar` there and put `(?P<bar>` instead of `(?P<date>`, or whatever foobar-like name you'd prefer.

feerrenrut commented on tests/unit/test_characterProcessing.py:
```diff
@@ -0,0 +1,40 @@
+		return m.group()
+
+	def test_replacement(self):
+		"""Test group replacement function"""
```
While this may test most of the paths (which is quite hard to determine at a glance), and is a good "happy path" test. I would like to see smaller examples that shows each of the paths. There are at least 6 paths through the `_replaceGroups` function, and many more combinations due to the internal loop. Please add an independent test for each path it can take. 

These tests act as documentation for what the expectations are for the function, so an unfamiliar developer can easily understand the different things the function can do.

Please also test the error case.

sthibaul replied:
Ok, I have separated out the tests.

feerrenrut commented on source/characterProcessing.py:
```diff
@@ -519,6 +519,33 @@ def __init__(self, locale):
 			log.error("Invalid complex symbol regular expression in locale %s: %s" % (locale, e))
 			raise LookupError
 
+	def _replaceGroups(self, m, string):
```
Rather than in the doc string, please add this as a type annotation.
```suggestion
	def _replaceGroups(self, m: re.Matchobject, string: str) -> str:

```

sthibaul replied:
Ok, done so

feerrenrut commented on tests/unit/test_characterProcessing.py:
```diff
@@ -0,0 +1,74 @@
+	def test_group_replacement(self):
+		"""Test that plain text gets properly replaced
+		"""
+		self.assertEqual(self._replace("1", r"(?P<foo>(\d))", "foo", "a"), "a")
```
Thanks for adding extra tests, at the moment it is hard to read the 'actual' and 'expected' values.There is a lot happening on these lines. Could you split them in the following way please?
```suggestion
		replaced = self._replace(
			string="1",
			pattern=r"(?P<foo>(\d))",  # name group 'foo' looking for a single digit.
			name="foo",
			replacement="a"
		)
		self.assertEqual(replaced, "a")

```

sthibaul replied:
Ok, done so

feerrenrut commented on tests/unit/test_characterProcessing.py:
```diff
@@ -0,0 +1,74 @@
+                    self._replace("1", r"(?P<foo>(\d))", "foo", r"\a")
+
+	def test_unterminated_escape(self):
+		"""Test that an escape at the end of string raises an error
```
Can you explain here why the error is raised (for other error cases too please)

sthibaul replied:
I don't understand what kind of explanation you would like. Something non-supported shall report an error, I don't see what else to explain? I guess I can say that an escape at the end of a string is meaningless since we don't know what is being escaped, is that really about it?

feerrenrut replied:
If the comment can answer "why?": Why is an escape at the end of the string an error?

feerrenrut replied:
Not only that, but the comment currently says "at the end of the string" but the string parameter has no escape. It would be clearer if it said replacement? The docs could show a slightly modified valid version.

sthibaul replied:
Ok, fixed so.

feerrenrut commented on tests/unit/test_characterProcessing.py:
```diff
@@ -0,0 +1,74 @@
+	def test_group_replacements(self):
+		"""Test that group references get properly replaced
+		"""
+		self.assertEqual(self._replace("1.2", r"(?P<foo>(\d)\.(\d))", "foo", r"\2 \1"), "2 1")
```
I think this would be clearer if it didn't use '2' and '1', since these match the number in the group. Perhaps something like:
```suggestion
		self.assertEqual(self._replace("foo.BAR", r"(?P<foo>([a-z])\.([A-Z))", "foo", r"\2>\1"), "BAR>foo")

```

sthibaul replied:
Ok, done so.

feerrenrut commented on source/characterProcessing.py:
```diff
@@ -519,6 +519,35 @@ def __init__(self, locale):
 
+	def _replaceGroups(self, m: re.Match, string: str) -> str:
+		"""Replace matching group references (\\1, \\2, ...) with the corresponding matched groups.
+		Also replace \\\\ with \\ and reject other escapes, for escaping coherency.
```
Is there a test for the `\\\\` case?

sthibaul replied:
I added one indeed, better check that quadruple backslash works.
Note that here in the doc string, `\\\\` is actually only two characters.

feerrenrut commented on tests/unit/test_characterProcessing.py:
```diff
@@ -0,0 +1,119 @@
+		replaced = self._replace(
+			string="bar.BAT",
+			pattern=r"(?P<foo>([a-z]*)\.([A-Z]*))",
+			name="foo",
```
I've just noticed that all these tests use name, when it is not strictly necessary to test the replace function. Why is this? I want to see tests for the custom code in the `_replaceGroups` code, which doesn't consider group name at all. The more simple the regex, the easier these tests will be to understand.

sthibaul replied:
Right, we can just put anonymous groups. The only exception if of course the test for multiple groups, we want to name the one we aim to replace.
Done so.

feerrenrut commented:
Hi @sthibaul can I request that you don't squash and force push every change? It makes it difficult to see what has changed since I last reviewed. Rest assured that we will squash merge to master, or if you want to clean the history of the branch please do so after the reviews are all complete.

feerrenrut commented on source/locale/fr/symbols.dic:
```diff
@@ -40,7 +40,7 @@ symbols:
 decimal point		# virgule décimale
 thousands separator		# point des mille
-dates .	\1 \2 \3 	all	always	# point de date
+dates .	\1 \2 \3 	all	norep	# point de date
```
@sthibaul Interesting to note that this dates replacement is not done in the English dictionary! 
@josephsl Will this modification cause a problem for translators?

sthibaul replied:
As I mentioned previously, this is a bug in the french translation, I'm discussing with the translator, so we get that included before this gets commited (without having to do the change, then, since it'll already be there).

feerrenrut replied:
I see. I'll add the blocked label to this PR so that it isn't merged until the translation change comes through. Please add a comment when that happens.

zstanecic commented:
Yes, it may cause problems.

 

From: Reef Turner <notifications@github.com> 
Sent: Wednesday, June 17, 2020 1:44 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nvaccess/nvda] symbols: support group references in replacements (#11116)

 

@feerrenrut commented on this pull request.

  _____  

In source/locale/fr/symbols.dic <https://github.com/nvaccess/nvda/pull/11116#discussion_r441482840> :

> @@ -40,7 +40,7 @@ symbols:
 
 
 
 decimal point         # virgule décimale
 
 thousands separator           # point des mille
 
-dates .       \1 \2 \3       all     always  # point de date
 
+dates .       \1 \2 \3       all     norep   # point de date
 

@sthibaul <https://github.com/sthibaul>  Interesting to note that this dates replacement is not done in the English dictionary!
@josephsl <https://github.com/josephsl>  Will this modification cause a problem for translators?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11116#pullrequestreview-432317828> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACVCDE7MZQWW2OVAGXHXRSLRXCT77ANCNFSM4MZ6EN7A> .  <https://github.com/notifications/beacon/ACVCDE33ZTQAIKFEQGQRSZDRXCT77A5CNFSM4MZ6EN7KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGODHCKLBA.gif>

sthibaul commented:
> Hi @sthibaul can I request that you don't squash and force push every change? It makes it difficult 

Ah, ok, sure, I thought github was providing an easy way to see the changes since last push (iirc gitlab does)

sthibaul committed:
Add a test for a reference to a missing group

CyrilleB79 commented:
Seems this PR is blocked due to French translation?

With NVDA 2020.2, reading "20/08/2020" with punctuation level set to all, we can hear the \1 \2 and \3 expressions. Maybe an unintentional modification in commit 56765 of assembla SVN translation repo.

@MichelSuch, since you are the author of this Assembla commit, would you mind to confirm my assumption and revert the incorrect date rule. Then inform here so that this PR may be unblocked. Thanks.

sthibaul commented:
> and revert the incorrect date rule

Well, I wouldn't say revert. We have a kind of chicken-and-egg here: we want to have a test that checks that translation rules do work, so we need a translation that has a rule using groups. French happens to do have one, even if for now it is bogus since the support is not there. But we can just unbreak that instead of reverting.

The question was rather about the "always" vs "norep", we are discussing the matter by mail (in French)

sthibaul commented:
But yes, you are right in pinging @MichelSuch , this PR is waiting for his feedback to my mails :)

CyrilleB79 commented:
I may not have understood the whole issue.
Anyway IMO we should not commit permanently test stuff that introduce a regression here.
You can always attach a modified test symbols.dic file to this PR if needed and describe the test procedure in this PR.

Feel free to add me in the French e-mail exchange to discuss further the French file issue. I also belong to French translation team, even if my contribution is very minimal with respect to Michel's or Rémy's ones.

sthibaul commented:
> Anyway IMO we should not commit permanently test stuff that introduce a regression here.

It was not actually meant to be commited, that was due to a misunderstanding.

> You can always attach a modified test symbols.dic file to this PR if needed and describe the test procedure in this PR.

The test is automatic, that's the whole point: to make sure that the code behave as expected. But I'm precisely exchanging with Michel to determine what is actually expected, because it does not seem that clear.

> Feel free to add me in the French e-mail exchange to discuss further the French file issue.

I don't know your email address. You can contact me on samuel.thibault@ens-lyon.org

sthibaul committed:
Fix example according to translator's update

sthibaul committed:
Revert to current fr status, translator has fixed this rule, pending merge

Also drop the test for now, will be commited after mergin the translator
update.

sthibaul commented:
Ok, we have sorted out the french translation question. Michel has commited reverting to not using groups, and using norep instead of always.

In this pull request I have dropped the fr_FR test, so that it can  be commited without waiting for the french translation update. 

When this pull request is merged, Michel will be able to update the french translation to use the group support, and then I'll submit another pull request to add a test for it.

CyrilleB79 commented:
I did not review the code. However I have some remarks:
* At first glance, the provided example does not seem realistic: why would a user replace the dot by space when the punctuation level is set to all? I would have expected something to be announced instead at this high level. The replacement rule could be for example (in French):
`dates .	\1 point \2 point \3 	all	always	# point de date`
  Maybe I have missed something, so please add an indication to let the reader know the context, i.e. why this rule was added in the French symbol file. @MichelSuch?
* Also the semantic of this rule has changed:
  * Before the rule was matching a dot in the date and replacing it:
  `dates .	 point 	all	norep	# point de date`
  * Now it matches the whole date and replace it with group substitution:
  `dates .	\1 \2 \3 	all	norep	# point de date`
  Thus I suggest renaming the rule from "dates ." to "date with ." in the example as well as in the translation repo French file when this file will have to be modified (after this PR is merged). Also the associated comment should specify "date avec point" (i.e. "date with dot") instead of "point de date" (i.e. "date dot")
* Regarding change log entry something more precise would help such as:
`Symbol pronunciation: Support grouping in complex symbol definition and group references in replacement rule`

sthibaul committed:
Drop unused imports

sthibaul committed:
Update dates naming

Commit:
Fix text

sthibaul committed:
Add some comments in the user guide as well

sthibaul committed:
Merge branch 'master' into groups

sthibaul committed:
drop bogus paragraph

sthibaul committed:
rephrase paragraph for user guide

sthibaul commented:
Any news on this?  @feerrenrut perhaps? The french translation question has been sorted out, we will update it after this gets commited.

feerrenrut commented:
Ok, thanks @sthibaul so I can take it that this is no longer blocked by the French translation and is ready for review and merging?

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into groups

feerrenrut committed:
update changes file for PR #11116
